engine/src/xcc.rs

```diff
@@ -0,0 +1,202 @@
+use aurora_engine_sdk::error::ReadU32Error;
+use aurora_engine_sdk::io::{StorageIntermediate, IO};
+use aurora_engine_sdk::promise::PromiseHandler;
+use aurora_engine_types::account_id::AccountId;
+use aurora_engine_types::parameters::{PromiseAction, PromiseBatchAction, PromiseCreateArgs};
+use aurora_engine_types::storage::{self, KeyPrefix};
+use aurora_engine_types::types::{Address, NearGas, Yocto, ZERO_YOCTO};
+use aurora_engine_types::Cow;
+use aurora_engine_types::Vec;
+use borsh::{BorshDeserialize, BorshSerialize};
+
+pub const ERR_NO_ROUTER_CODE: &str = "ERR_MISSING_XCC_BYTECODE";
+pub const ERR_CORRUPTED_STORAGE: &str = "ERR_CORRUPTED_XCC_STORAGE";
+pub const ERR_INVALID_ACCOUNT: &str = "ERR_INVALID_XCC_ACCOUNT";
+pub const ERR_ATTACHED_NEAR: &str = "ERR_ATTACHED_XCC_NEAR";
+pub const VERSION_KEY: &[u8] = b"version";
+pub const CODE_KEY: &[u8] = b"router_code";
+pub const VERSION_UPDATE_GAS: NearGas = NearGas::new(5_000_000_000_000);
+pub const INITIALIZE_GAS: NearGas = NearGas::new(5_000_000_000_000);
+/// Amount of NEAR needed to cover storage for a router contract.
+pub const STORAGE_AMOUNT: Yocto = Yocto::new(2_000_000_000_000_000_000_000_000);
+
+/// Type wrapper for version of router contracts.
+#[derive(
+    Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, BorshDeserialize, BorshSerialize,
+)]
+pub struct CodeVersion(pub u32);
+
+impl CodeVersion {
+    pub fn increment(self) -> Self {
+        Self(self.0 + 1)
+    }
+}
+
+/// Type wrapper for router bytecode.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct RouterCode<'a>(pub Cow<'a, [u8]>);
+
+impl<'a> RouterCode<'a> {
+    pub fn new(bytes: Vec<u8>) -> Self {
+        Self(Cow::Owned(bytes))
+    }
+
+    pub fn borrowed(bytes: &'a [u8]) -> Self {
+        Self(Cow::Borrowed(bytes))
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshDeserialize, BorshSerialize)]
+pub struct AddressVersionUpdateArgs {
+    pub address: Address,
+    pub version: CodeVersion,
+}
+
+pub fn handle_precompile_promise<I, P>(
+    io: &I,
+    handler: &mut P,
+    promise: PromiseCreateArgs,
+    current_account_id: &AccountId,
+) where
+    P: PromiseHandler,
+    I: IO,
+{
+    let target_account: &str = promise.target_account_id.as_ref();
+    let sender = Address::decode(&target_account[0..40]).expect(ERR_INVALID_ACCOUNT);
+
+    // Confirm target_account is of the form `{address}.{aurora}`
+    // Address prefix parsed above, so only need to check `.{aurora}`
+    assert_eq!(&target_account[40..41], ".", "{}", ERR_INVALID_ACCOUNT);
+    assert_eq!(
+        &target_account[41..],
```

Comment:
I think this is exactly the case where we do not want an ending index. The assert is to check that whole rest of the string matches the current account id. It could be done by checking the total length and then including the length bound in this check, but I don't see what security that adds over how it is written currently.

Reply:
Sure, thats fair. I'll reconsider limiting that then.